File: lang/JOpalIF/Opal.f
\ JForth Language Interface for OpalVision's Opal.Library
\
\ Version 1.0 - 31 December 1992
\
\ By Marlin Schwanke

include? OpalScreen ji:opal/opallib.j

ANEW TASK-OPAL

\ JForth Amiga shared library words

:Library opal
: opal? ( -- ) opal_NAME opal_LIB lib? ;
: -opal ( -- ) opal_LIB -lib ;

\ Some useful macros

\ Set 24 bit pen value
: SetPen24 ( OpalScreen, Red8, Green8, Blue8 -- , Store 24 bitplane pen )
   >r >r                ( Now have OS R8 )
   over ..! OS_Pen_R     ( Store red pen )
   r> over ..! OS_Pen_G  ( Store green pen )
   r> swap ..! OS_Pen_B  ( Store blue pen
;

\ Set 15 bit pen value
: SetPen15 ( OpalScreen Red5 Green5 Blue5 -- , Store 15 bitplane pen )
   31 and >r 31 and swap 31 and        ( Now have OS 000GGGGG 000RRRRR )
   2 shift over  24 and -3 shift or    ( Now have OS 0RRRRRGG )
   swap 7 and 5 shift r> or            ( Now have OS 0RRRRRGG GGGBBBBB )
   2 pick ..! OS_Pen_G                  ( Store green pen )
   swap ..! OS_Pen_R                    ( Store red pen )
;

\ Set 8 bit pallette mapped pen value
: SetPen8P ( OpalScreen Pen8 -- , Store 8 bitplane pen )
   swap ..! OS_Pen_R
;

\ Set 8 bit pen value
: SetPen8
   3 and >r 7 and swap 7 and           ( Now have OS G3 R3 )
   5 shift swap 2 shift or r> or       ( Now have OS RRRGGGBB )
   swap ..!  OS_Pen_R                   ( Store Pen_R )
;

\ Get 24 bit pen values
: GetPen24 ( OpalScreen -- Red8 Green8 Blue8 , Get the current 24 bitplane pen )
   dup ..@ OS_Pen_R        ( Now have OS Pen_R )
   swap dup ..@ OS_Pen_G   ( Now have Pen_R OS Pen_G )
   swap ..@ OS_Pen_B       ( Now have Pen_R Pen_G Pen_B )
;

\ Get 15 bit pen values
: GetPen15 ( OpalScreen -- Red5 Green5 Blue5 , Get the current 15 bitplane pen )
   dup ..@ OS_Pen_R                    ( Now have OS 0RRRRRGG )
   dup -2 shift                        ( Now have OS 0RRRRRGG 000RRRRR )
   swap 3 and 3 shift                  ( Now have OS 000RRRRR 000GG000 )
   rot ..@ OS_Pen_G                    ( Now have 000RRRRR 000GG000 GGGBBBBB )
   swap over 224 and -5 shift or       ( Now have 000RRRRR GGGBBBBB 000GGGGG )
   swap 31 and                         ( Now have 000RRRRR 000GGGGG 000BBBBB )
;

\ Get 8 bit palette mapped pen value
: GetPen8P ( OpalScreen -- Pen8 , Get the current 8 bitplane pen )
   ..@ OS_Pen_R
;

\ Get 8 bit pen values
: GetPen8 ( OpalScreen -- Red3 Green3 Blue2 )
   ..@ OS_Pen_R
   dup -5 shift            ( Now have OS 00000RRR )
   over 28 and -2 shift    ( Now have OS 00000RRR 00000GGG )
   rot 3 and               ( Now have 00000RRR 00000GGG 000000BB )
;

\ Get 24 bit color values
: GetCol24 ( OpalScreen -- Red8, Green8 , Blue 8 , Get ReadPixel24 24 bitplane color values )
   dup ..@ OS_Red       ( Now have OS Red8 )
   over ..@ OS_Green    ( Now have OS Red8 Green8 )
   rot ..@ OS_Blue      ( Now have Red8 Green8 Blue8 )
;

\ Get 15 bit color values
: GetCol15 ( OpalScreen -- Red5 Green5 Blue5 , Get ReadPixel24 15 bitplane color values )
   dup ..@ OS_Red                   ( Now have OS 0RRRRRGGG )
   dup -2 shift                     ( Now have OS 0RRRRRGG 000RRRRR )
   swap 3 and 3 shift               ( Now have OS 000RRRRR 000GG000 )
   rot ..@ OS_Green                 ( Now have 000RRRRR 00000GG GGGBBBBB )
   swap over 224 and -5 shift or    ( Now have 000RRRRR GGGBBBBB 000GGGGG )
   swap 31 and                      ( Now have 000RRRRR 000GGGGG 000BBBBB )
;

\ Get 8 bit palette mapped color value
: GetCol8P ( OpalScreen -- Col8 , Get ReadPixel24 8 bitplane color value )
   ..@ OS_Red
;

\ Get 8 bit color values
: GetCol8 ( OpalScreen -- Red3 Green3 Blue2 )
   ..@ OS_Red
   dup -5 shift            ( Now have OS 00000RRR )
   over 28 and -2 shift    ( Now have OS 00000RRR 00000GGG )
   rot 3 and               ( Now have 00000RRR 00000GGG 000000BB )
;

\ Playfield & Priority stencil pens
: SetPFPen ( Opalscreen State -- Set playfield pen to state )
   swap ..! OS_Pen_R
;

: SetPRPen ( OpalScreen State -- Set priority pen to state )
   swap ..! OS_Pen_R
;

\ Opal Library Calls

\ Note that all addresses are JForth relative going into and coming out of these
\ routines.  The code for each call handles the necessary transformations.  Also
\ note that all filenames must be 0 terminated 'C' style strings.

: OpenScreen24()  ( Modes -- Screen/Null )
\ Allocates all resources and displays an OpalVision screen.
   call Opal_LIB OpenScreen24 if>rel
;

: CloseScreen24() ( -- )
\ Stop current display and free resources.
   callvoid Opal_LIB CloseScreen24
;

: WritePixel24() ( Screen X Y -- Result )
\ Write pixel into an OpalScreen.
   call>abs Opal_LIB WritePixel24
;

: ReadPixel24() ( Screen X Y -- Result )
\ Returns colour information for a given pixel.
   call>abs Opal_LIB ReadPixel24
;

: ClearScreen24() ( Screen -- )
\ Clears all bitplanes in a screen.
   callvoid>abs Opal_LIB ClearScreen24
;

: ILBMtoOV() ( Screen ILBMData SourceWidth Lines TopLine Planes -- )
\ Convert interleaved bitmap to OpalVision format.
   callvoid>abs Opal_LIB ILBMtoOV
;

: UpdateDelay24() ( Frames -- )
\ Sets the delay between consecutive frame buffer updates.
   callvoid Opal_LIB UpdateDelay24
;

: Refresh24() ( -- )
\ Refreshes the frame buffer.
   callvoid Opal_LIB Refresh24
;

: SetDisplayBottom24() ( BottomLine -- Result )
\ Sets the lower limit of the OpalVision screen.
   call Opal_LIB SetDisplayBottom24
;

: ClearDisplayBottom24() ( -- )
\ Clears the OpalVision display bottom setting.
   callvoid Opal_LIB ClearDisplayBottom24
;

: SetSprite24() ( SpriteData SpriteNumber -- )
\ Allows Amiga sprites to be displayed over OpalVision graphics.
   callvoid>abs Opal_LIB SetSprite24
;

: AmigaPriority() ( -- )
\ Gives Amiga graphics priority over OpalVision display.
   callvoid Opal_LIB AmigaPriority
;

: OVPriority() ( -- )
\ Give OpalVision graphics priority over the Amiga graphics.
   callvoid Opal_LIB OVPriority
;

: DualDisplay24() ( -- )
\ Sets up an Amiga/OpalVision dual display.
   callvoid Opal_LIB DualDisplay24
;

: SingleDisplay24() ( -- )
\ Sets up an Amiga/OpalVision single display.
   callvoid Opal_LIB SingleDisplay24
;

: AppendCopper24() ( CopperArray -- )
\ Attaches user copper lists to existing display copper lists.
   callvoid>abs Opal_LIB AppendCopper24
;

: RectFill24() ( OScrn Left Top Bottom Right -- )
\ Draws a solid rectangle.
   callvoid>abs Opal_LIB RectFill24
;

: UpdateCoPro24() ( -- )
\ Writes CoPro list for current the display screen to Video coprocessor.
   callvoid Opal_LIB UpdateCoPro24
;

: SetControlBit24() ( FrameNumber BitNumber State -- )
\ Modifies a bit in the control line register.
   callvoid Opal_LIB SetControlBit24
;

: PaletteMap24() ( PaletteMap -- )
\ Enable/Disable palette mapping.
   callvoid Opal_LIB PaletteMap24
;

: UpdatePalette24() ( -- )
\ Loads all 256 entries of Red, Green, and Blue values in the OpalScreen
\ structure into the OpalVision palette registers.
   callvoid Opal_LIB UpdatePalette24
;

: Scroll24() ( DeltaX DeltaY -- )
\ Scrolls currently displayed OpalVision image.
   callvoid Opal_LIB Scroll24
;

: LoadIff24() ( OpalScreen 0Filename Flags -- OpalScreen/Error )
\ Loads an image file.
   >r >abs swap            ( fn scr scr )
   if>abs                  ( fn scr/0 )
   swap r>                 ( scr/0 fn fl )
   call Opal_LIB LoadIff24 ( scr/error )
   dup OL_ERR_MAXERR >=    ( scr/error )
   if >rel then            ( scr/error )
;

: LoadImage24() ( OpalScreen 0Filename Flags -- OpalScreen/Error )
\ Alias for LoadIff24().
   LoadIFF24()
;

: SetScreen24() ( OpalScreen -- )
\ Fills screen with a specified colour.
   callvoid>abs Opal_LIB SetScreen24
;

: SaveIFF24() ( OpalScreen 0Filename ChunkFunction Flags -- Result )
\ Save an OpalScreen as an IFF file.
   call>abs Opal_LIB SaveIFF24
;

: CreateScreen24() ( ScreenModes Width Height -- OpalScreen/Null )
\ Create an arbitrarily sized virtual OpalScreen.
   call Opal_LIB CreateScreen24 if>rel
;

: FreeScreen24() ( OpalScreen -- )
\ Frees a virtual OpalScreen.
   callvoid>abs Opal_LIB FreeScreen24
;

: UpdateRegs24() ( -- )
\ Updates the hardware registers for the current display screen.
   callvoid Opal_LIB UpdateRegs24
;

: SetLoadAddress24() ( -- )
\ Updates the OpalVision load address register.
   callvoid Opal_LIB SetLoadAddress24
;

: RGBtoOV() ( OpalScreen RGBPlanes Top Left Width height -- )
\ Convert three planes of RGB to OpalVision bitplane data.
   callvoid>abs Opal_LIB RGBtoOV
;

: ActiveScreen24() ( -- OpalScreen/Null )
\ Provides a pointer to the currently displayed OpalVision screen.
   call Opal_LIB ActiveScreen24 if>rel
;

: FadeIn24() ( Time -- )
\ Fades display in from black.
   callvoid Opal_LIB FadeIn24
;

: FadeOut24() ( Time -- )
\ Fade display to black.
   callvoid Opal_LIB FadeOut24
;

: ClearQuick24() ( -- )
\ Clears frame buffer memory.
   callvoid Opal_LIB ClearQuick24
;

: WriteThumbnail24() ( OpalScreen File -- Result )
\ Writes an IFF thumb-nail chunk into a file.
   callvoid>abs Opal_LIB WriteThumbnail24
;

: SetRGB24() ( Entry Red Green Blue -- )
\ Updates a single palette entry to the OpalVision palette registers.
   callvoid Opal_LIB SetRGB24
;

: DrawLine24() ( OpalScreen X1 Y1 X2 Y2 -- )
\ Draws a line into an OpalScreen.
   callvoid>abs Opal_LIB DrawLine24
;

: StopUpdate24() ( -- )
\ Stops updates to the frame buffer memory.
   callvoid Opal_LIB StopUpdate24
;

: WritePFPixel24() ( OpalScreen X Y -- )
\ Set or clear a pixel in the playfield stencil.
   callvoid>abs Opal_LIB WritePFPixel24
;

: WritePRPixel24() ( OpalScreen X Y -- )
\ Set or clear a pixel in the priority stencil.
   callvoid>abs Opal_LIB WritePRPixel24
;

: OVtoRGB() ( OpalScreen RGBPlanes Top Left Width Height -- )
\ Converts OpalVision bitplane data to three planes of RGB
   callvoid>abs Opal_LIB OVtoRGB
;

: OVtoILBM() ( OpalScreen ILBMData DestWidth Lines TopLine -- )
\ Convert OpalVision bit planes to interleaved bitmap format.
   callvoid>abs Opal_LIB OVtoILBM
;

: UpdateAll24() ( -- )
\ Resets the internal update stucture so all required banks are updated.
   callvoid Opal_LIB UpdateAll24
;

: UpdatePFStencil24() ( -- )
\ Updates playfield stencil at highest possible rate.
   callvoid Opal_LIB UpdatePFStencil24
;

: EnablePRStencil24() ( -- )
\ Enables the use of the priority stencil in dual display mode.
   callvoid Opal_LIB EnablePRStencil24
;

: DisablePRStencil24() ( -- )
\ Disables the use of the priority stencil in dual display mode.
   callvoid Opal_LIB DisablePRStencil24
;

: ClearPRStencil24() ( OpalScreen -- )
\ Clears the priority stencil of the specified screen.
   callvoid>abs Opal_LIB ClearPRStencil24
;

: SetPRStencil24() ( OpalScreen -- )
\ Sets the priority stencil of the specified screen.
   callvoid>abs Opal_LIB SetPRStencil24
;

: DisplayFrame24() ( Frame -- )
\ Sets the currently displayed frame within the frame buffer memory.
   callvoid Opal_LIB DisplayFrame24
;

: WriteFrame24() ( Frame -- )
\ Sets the current frame to be written within the frame buffer memory.
   callvoid Opal_LIB WriteFrame24
;

: BitPlanetoOV() ( OpalScreen ScrPlanes ScrWidth Lines TopLine ScrDepth -- )
\ Convert standard bitplane data to OpalVision format.
   callvoid>abs Opal_LIB BitPlanetoOV
;

: SetCoPro24() ( InstructionNumber Instruction -- )
\ Modifies a single instruction in the CoPro list.
   callvoid Opal_LIB SetCoPro24
;

: RegWait24() ( -- )
\ Wait for register update to complete.
   callvoid Opal_LIB RegWait24
;

: DualPlayField24 ( -- )
\ Sets up an OpalVision 24 bit dual playfield.
   callvoid Opal_LIB DualPlayField24
;

: SinglePlayField24 ( -- )
\ Set up an Amiga or OpalVision single playfield.
   callvoid Opal_LIB SinglePlayField24
;

: ClearPFStencil24() ( OpalScreen -- )
\ Clears the playfield stencil of the specified screen.
   callvoid>abs Opal_LIB ClearPFStencil24
;

: SetPFStencil24() ( OpalScreen -- )
\ Sets the playfield stencil of the specified screen.
   callvoid>abs Opal_LIB SetPFStencil24
;

: ReadPRPixel24() ( OpalScreen X Y --  Result )
\ Returns the state of a given priority stencil pixel.
   call>abs Opal_LIB ReadPRPixel24
;

: ReadPFPixel24() ( OpalScreen X Y --  Result )
\ Returns the state of a given playfield stencil pixel.
   call>abs Opal_LIB ReadPFPixel24
;

: OVtoBitPlane() ( OpalScreen BitPlanes DestWidth Lines TopLine -- )
\ Convert OpalVision bit plane data to standard bitplanes.
   callvoid>abs Opal_LIB OVtoBitPlane
;

: FreezeFrame24() ( Freeze -- )
\ Freezes the currently displayed screen.
   callvoid Opal_LIB FreezeFrame24
;

: LowMemUpdate24() ( OpalScreen Frame -- OpalScreen/Error )
   call>abs Opal_LIB LowMemUpdate24
   dup OL_ERR_MAXERR >=    ( scr/error )
   if >rel then            ( scr/error )
;

: DisplayThumbnail24() ( Screen 0Filename X Y -- Result )
\ Displays a file's thumbnail.
   call>abs Opal_LIB DisplayThumbnail24
;

: Config24() ( -- Result )
\ Returns the OpalVision hardware configuration.
   call Opal_LIB Config24
;

: AutoSync24() ( Sync -- )
\ Enables auto horizontal synchronisation.
   callvoid Opal_LIB AutoSync24
;

: DrawEllipse24() ( OpalScreen CX CY A B -- )
\ Draw an ellipse of given dimensions.
   callvoid>abs Opal_LIB DrawEllipse24
;

: DrawCircle24() ( opalScreen centerx centery radius -- , Draw circle )
\ An alias for DrawEllipse24() for drawing circles.
   dup DrawEllipse24()
;

: LatchDisplay24 ( Latch -- )
\ Locks OpalVision display.
   callvoid Opal_LIB LatchDisplay24
;

: SetHires24() ( Topline Lines -- )
\ Enable a hires display for a section of the screen.
   callvoid Opal_LIB SetHires24
;

: SetLores24() ( Topline Lines -- )
\ Enable a lores display for a section of the screen.
   callvoid Opal_LIB SetLores24
;

: DownLoadFrame24() ( OpalScreen X Y W H -- )
\ ??
   callvoid>abs Opal_LIB DownLoadFrame24
;

: SaveJPEG24() ( OpalScreen 0FileName Flags Quality -- Error )
\ Save an OpalScreen as a JPEG JFIF file.
   call>abs Opal_LIB SaveJPEG24
;

: LowMem2Update24() ( Screen Frame -- OpalScreen/Error )
\ Low chip ram usage OpalVision update.
   call>abs Opal_LIB LowMem2Update24
   dup OL_ERR_MAXERR >=       ( scr/error )
   if >rel then               ( scr/error )
;

: LowMemRGB24() ( ScreenModes Frame Width Height Modulo RGBPlanes -- )
\ Low chip ram usage OpalVision update from an RGB array.
   callvoid>abs Opal_LIB LowMemRGB24
   dup OL_ERR_MAXERR >=       ( scr/error )
   if >rel then               ( scr/error )
;

File: lang/hspasov4/Opal.pas
{--------------------------------------------------------------------------

                         OPALVISION LIBRARY UNIT

                   ( HighSpeed Pascal for the Amiga )

                   Programmed by Kelly J. Petlig 1992

                 Copyright (c) 1992 All rights reserved

  Version : Date (mm/dd/yy) : Comment
  -----------------------------------
    1.00 : 12/27/92 : First version
    1.01 : 12/30/92 : Fixed bug (called procedures as functions)
    1.02 : 12/31/92 : Added OpalRequestor procedure
    1.03 : 01/11/93 : CloseScreen24 now is a procedure, not a function
    1.04 : 01/13/93 : Bug in _fd file caused problem with LowMemUpdate24

--------------------------------------------------------------------------}

Unit Opal;

INTERFACE

Uses Exec, Intuition;

Const

  MAXCOPROINS   = 290;     { Number of CoPro instructions }

          { Screen flags }

  HIRES24       = $0001;   { High resolution screen. }
  ILACE24       = $0002;   { Interlaced screen. }
  OVERSCAN24    = $0004;   { Overscan screen. }
  NTSC24        = $0008;   { NTSC Screen - Not user definable }
  CLOSEABLE24   = $0010;   { Screen is closeable. }
  PLANES8       = $0020;   { Screen has 8 bitplanes. }
  PLANES15      = $0040;   { Screen has 15 bitplanes. }
  CONTROLONLY24 = $2000;   { Used for updating control bits only }
  PALMAP24      = $4000;   { Screen is in palette mapped mode }
  INCHIP24      = $8000;   { In chip ram - Not user definable }

  FLAGSMASK24 = $6077;

          { LoadImage24 Flags }

  CONVERT24       = 1;  { Force conversion of palette mapped to 24 bit }
  KEEPRES24       = 2;  { Keep the current screen resolution }
  LOADMASK24      = 4;  { Load mask plane if it exists }
  VIRTUALSCREEN24 = 8;  { Load complete image into fast ram }

          { SaveIFF24 Flags }

  OVFASTFORMAT = 1;     { Save as opalvision fast format }
  NOTHUMBNAIL  = 4;     { Inhibit thumbnail chunk }
  SAVEMASK24   = 8;     { Save MaskPlane with image }

          { Config Flags }

  OVCF_OPALVISION = 1;      { Display board is an OpalVision }
  OVCF_COLORBURST = 2;      { Display board is a ColorBurst }

          { Opal Requestor Flags }

  NO_INFO   = 1;        { Exclude files ending in .info }
  LASTPATH  = 2;        { Use Last selected path as current }

  OR_ERR_OUTOFMEM = 1;
  OR_ERR_INUSE    = 2;

  OPALREQ_HEIGHT  = 345;    { The height of the Requester }

          { Coprocessor bits. }

  VIDMODE0     = $01;    { Video control bit 1 (S0) }
  VIDMODE1     = $02;    { Video control bit 1 (S1) }
  DISPLAYBANK2 = $04;    { Select display bank 2 }
  HIRESDISP    = $08;    { Enable hi-res display }
  DUALDISPLAY  = $10;    { Select dual display mode (active low) }
  OVPRI        = $20;    { Set OpalVision priority }
  PRISTENCIL   = $40;    { Enable priority stencil }
  ADDLOAD      = $80;    { Address load bit. Active low }

  ADDLOAD_B      = 7;
  PRISTENCIL_B   = 6;
  OVPRI_B        = 5;
  DUALDISPLAY_B  = 4;
  HIRESDISP_B    = 3;
  DISPLAYBANK2_B = 2;
  VIDMODE1_B     = 1;
  VIDMODE0_B     = 0;

          { Control line bits }

  VALID0        = $00001;
  VALID1        = $00002;
  VALID2        = $00004;
  VALID3        = $00008;
  WREN          = $00010;
  COL_COPRO     = $00020;
  AUTO          = $00040;
  DUALPLAYFIELD = $00080;
  FIELD         = $00100;
  AUTOFIELD     = $00200;
  DISPLAYLATCH  = $00400;
  FRAMEGRAB     = $00800;
  RWR1          = $01000;
  RWR2          = $02000;
  GWR1          = $04000;
  GWR2          = $08000;
  BWR1          = $10000;
  BWR2          = $20000;
  VLSIPROG      = $40000;
  FREEZEFRAME   = $80000;

  VALID0_B        =  0;
  VALID1_B        =  1;
  VALID2_B        =  2;
  VALID3_B        =  3;
  WREN_B          =  4;
  COL_COPRO_B     =  5;
  AUTO_B          =  6;
  DUALPLAYFIELD_B =  7;
  FIELD_B         =  8;
  AUTOFIELD_B     =  9;
  DISPLAYLATCH_B  = 10;
  FRAMEGRAB_B     = 11;
  RWR1_B          = 12;
  RWR2_B          = 13;
  GWR1_B          = 14;
  GWR2_B          = 15;
  BWR1_B          = 16;
  BWR2_B          = 17;
  VLSIPROG_B      = 18;
  FREEZEFRAME_B   = 19;

  NUMCONTROLBITS  = 20;
  VALIDCODE       =  5;

          { Error return codes }

  OL_ERR_OUTOFMEM      =  1;
  OL_ERR_OPENFILE      =  2;
  OL_ERR_NOTIFF        =  3;
  OL_ERR_FORMATUNKNOWN =  3;
  OL_ERR_NOTILBM       =  4;
  OL_ERR_FILEREAD      =  5;
  OL_ERR_FILEWRITE     =  6;
  OL_ERR_BADIFF        =  7;
  OL_ERR_CANTCLOSE     =  8;
  OL_ERR_OPENSCREEN    =  9;
  OL_ERR_NOTHUMBNAIL   = 10;
  OL_ERR_BADJPEG       = 11;
  OL_ERR_UNSUPPORTED   = 12;
  OL_ERR_CTRLC         = 13;
  OL_ERR_MAXERR        = 40;

Type
  tOpalScreen = Record
    Width : Integer;
    Height : Integer;
    Depth : Integer;            { Is this actually Depth / 2 ? }
    ClipX1,ClipY1 : Integer;
    ClipX2,ClipY2 : Integer;
    BytesPerLine : Integer;
    Flags : Word;
    RelX : Integer;
    RelY : Integer;
    UserPort : pMsgPort;
    MaxFrames : Integer;
    VStart : Integer;
    CoProOffset : Integer;
    LastWait : Integer;
    LastCoProIns : Word;
    BitPlanes : Array [0..23] of pShortInt; { 24 }
    MaskPlane : pShortInt;
    AddressReg : LongInt;
    UpdateDelay : Byte;
    PalLoadAddress : Byte;
    PixelReadMask : Byte;
    CommandReg : Byte;
    Palette : Array [0..767] of Byte; { 3 * 256 }
    Pen_R : Byte;
    Pen_G : Byte;
    Pen_B : Byte;
    Red : Byte;
    Green : Byte;
    Blue : Byte;
    CoProData : Array [0..289] of Byte; { MAXCOPROINS }
    Modulo : Integer;
    Reserved : Array [0..37] of Byte; { 38 }

  {$Ifdef OPAL_PRIVATE }
    CopList_Cycle : Array [0..11] of LongInt; { 12 }
    Update_Cycles : Byte;
    Pad : Byte;
  {$endif }

    end;

  pOpalScreen = ^tOpalScreen;

  tOpalReqBase = Record
          OR_Lib : tLibrary;
          OR_SegList : LongInt;
          end;

  tOpalReq = Record
          TopEdge : Word;         { Top Line of requester }
          Hail : pShortInt;       { Hailing text }
          File_ : pShortInt;      { Filename buffer (>=31 chars) }
          Dir : pShortInt;        { Directory name. }
          Extension : pShortInt;  { File extension to include }
          Window : pWindow;       { Window to display requester. }
          OScrn : pOpalScreen;    { OpalScreen to display req. }
          Pointer : pWord;        { Sprite mouse pointer }
          OKHit : Integer;        { TRUE if OK gadget hit }
          NeedRefresh : Integer;  { OpalScreen needs a refresh }
          Flags : LongInt;        { See Below }
          BackPen : Integer;      { Pen # to use for BG rendering }
          PrimaryPen : Integer;   { Pen # for primary rendering }
          SecondaryPen : Integer; { Pen # for secondary rendering }
          end;

pOpalReq = ^tOpalReq;

Var
  OpalBase: pLibrary;
  OpalReqBase : pOpalReq;

Function OpenScreen24 (Modes: LongInt): pOpalScreen;
Procedure CloseScreen24;
Function WritePixel24
                 (Screen: pOpalScreen;
                  x,
                  y: LongInt): LongInt;

Function ReadPixel24
                 (Screen: pOpalScreen;
                  x,
                  y: LongInt): LongInt;

Procedure ClearScreen24 (Screen: pOpalScreen);
Procedure ILBMtoOV
                 (Screen : pOpalScreen;
                  ILBMData: Pointer;
                  SourceWidth,
                  Lines,
                  TopLine,
                  Planes: LongInt);

Procedure UpdateDelay24 (Frames: LongInt);
Procedure Refresh24;
Function SetDisplayBottom24 (BottomLine: LongInt): LongInt;
Procedure ClearDisplayBottom24;
Procedure SetSprite24
                 (SpriteData: Pointer;
                  SpriteNum: LongInt);

Procedure AmigaPriority;
Procedure OVPriority;
Procedure DualDisplay24;
Procedure SingleDisplay24;
Procedure AppendCopper24 (CopperArray: Pointer);
Procedure RectFill24
                 (Screen: pOpalScreen;
                  x1,
                  y1,
                  x2,
                  y2: LongInt);

Procedure UpdateCoPro24;
Procedure SetControlBit24
                 (List,
                  Bit,
                  State: LongInt);

Procedure PaletteMap24 (State: LongInt);
Procedure UpdatePalette24;
Procedure Scroll24
                 (Deltax,
                  Deltay: LongInt);

Function LoadImage24
                 (Screen: pOpalScreen;
                  FileName: pCString;
                  Flags: LongInt): LongInt;

Procedure SetScreen24 (Screen: pOpalScreen);
Function SaveIFF24
                 (Screen: pOpalScreen;
                  FileName: pCString;
                  ChunkFunc: Pointer;
                  Flags: LongInt): LongInt;

Function CreateScreen24
                 (ScreenModes,
                  Width,
                  Height: LongInt): pOpalScreen;

Procedure FreeScreen24 (Screen: pOpalScreen);
Procedure UpdateRegs24;
Procedure SetLoadAddress24;
Procedure RGBtoOV
                 (Screen: pOpalScreen;
                  RGBData: Pointer;
                  x,
                  y,
                  w,
                  h: LongInt);

Function ActiveScreen24: pOpalScreen;
Procedure FadeIn24 (HundredthsSecs: LongInt);  { Must be > 1 }
Procedure FadeOut24 (HundredthsSecs: LongInt);   { Must be > 1 }
Procedure ClearQuick24;
Function WriteThumbnail24
                 (Screen: pOpalScreen;
                  File_: Pointer): LongInt;

Procedure SetRGB24
                 (Entry,
                  R,
                  G,
                  B: LongInt);

Procedure DrawLine24
                 (Screen: pOpalScreen;
                  x1,
                  y1,
                  x2,
                  y2: LongInt);

Procedure StopUpdate24;
Function WritePFPixel24
                 (Screen: pOpalScreen;
                  x,
                  y: LongInt): LongInt;

Function WritePRPixel24
                 (Screen: pOpalScreen;
                  x,
                  y: LongInt): LongInt;

Function OVtoRGB
                 (Screen: pOpalScreen;
                  RGBData: Pointer;
                  x,
                  y,
                  w,
                  h: LongInt): LongInt;

Procedure OVtoILBM
                 (Screen: pOpalScreen;
                  ILBMData: Pointer;
                  DestWidth,
                  Lines,
                  TopLine: LongInt);

Procedure UpdateAll24;
Procedure UpdatePFStencil24;
Procedure EnablePRStencil24;
Procedure DisablePRStencil24;
Procedure ClearPRStencil24 (Screen: pOpalScreen);
Procedure SetPRStencil24 (Screen: pOpalScreen);
Procedure DisplayFrame24 (Frame: LongInt);
Procedure WriteFrame24 (Frame: LongInt);
Procedure BitPlanetoOV
                 (Screen: pOpalScreen;
                  SrcPlanes: Pointer;
                  BytesPerLine,
                  Lines,
                  TopLine,
                  Depth: LongInt);

Procedure SetCoPro24
                 (Line,
                  Instruction: LongInt);

Procedure RegWait24;
Procedure DualPlayField24;
Procedure SinglePlayField24;
Procedure ClearPFStencil24 (Screen: pOpalScreen);
Procedure SetPFStencil24 (Screen: pOpalScreen);
Function ReadPRPixel24
                 (Screen: pOpalScreen;
                  x,
                  y: LongInt): LongInt;

Function ReadPFPixel24
                 (Screen: pOpalScreen;
                  x,
                  y: LongInt): LongInt;

Procedure OVtoBitPlane
                 (Screen: pOpalScreen;
                  DestPlanes: Pointer;
                  DestWidth,
                  Lines,
                  TopLine: LongInt);

Procedure FreezeFrame24 (Freeze: LongInt);
Function LowMemUpdate24
                 (Screen: pOpalScreen;
                  Frame: LongInt): pOpalScreen;
Function DisplayThumbnail24
                 (Screen: pOpalScreen;
                  FileName: pCString;
                  x,
                  y: LongInt): LongInt;

Function Config24: LongInt;
Procedure AutoSync24 (Sync: LongInt);
Procedure DrawEllipse24
                 (Screen: pOpalScreen;
                  Cx,
                  Cy,
                  a,
                  b: LongInt);

Procedure LatchDisplay24 (Latch: LongInt);
Procedure SetHires24
                 (TopLine,
                  Lines: LongInt);

Procedure SetLores24
                 (TopLine,
                  Lines: LongInt);

Function DownLoadFrame24
                 (Screen: pOpalScreen;
                  x,
                  y,
                  w,
                  h: LongInt): LongInt;

Function SaveJPEG24
                 (Screen: pOpalScreen;
                  FileName: pCString;
                  Flags,
                  Quality: LongInt): LongInt;

Function LowMem2Update24
                 (Screen: pOpalScreen;
                  Frame: LongInt): pOpalScreen;

Function LowMemRGB24
                 (Screen: pOpalScreen;
                  Frame,
                  Width,
                  Height,
                  Modulo: LongInt;
                  RGBPlanes: Pointer): pOpalScreen;

Procedure OpalRequestor (OpalReq : pOpalReq);
